File: cpp/leetcode/127.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordList) {
        queue<string> q;
        q.push(beginWord);
        int cnts = 1;
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                string word = q.front(); q.pop();
                if(word == endWord) return cnts;
                wordList.erase(word);
                for(int j = 0; j < word.size(); j++){
                    char ch = word[j];
                    for(int k = 0; k < 26; k++){
                        word[j] = 'a' + k;
                        if(wordList.find(word) != wordList.end()){
                            q.push(word); wordList.erase(word);
                        }
                    }
                    word[j] = ch;
                }
            }
            cnts++;
        }
        return 0;
    }
};

File: python/leetcode/127.py
class Solution:
    def ladderLength(self, beginWord, endWord, WordList):
        from collections import deque
        from string import ascii_lowercase
        WordList.add(endWord)
        n = len(beginWord)
        queue = deque([(beginWord, 1)])
        while queue:
            word, res = queue.popleft()
            if word == endWord: return res
            for i in xrange(n):
                for ch in ascii_lowercase:
                    if word[i] == ch: continue
                    new = word[:i] + ch + word[i+1:]
                    if new in WordList:
                        queue.append((new, res + 1))
                        WordList.remove(new)
        
        return 0